### PR TITLE
Allow user initiated navigation when the tour is active

### PR DESCRIPTION
**Pull request by mivanyna**

### CODE REVIEW

**Comment by hakimio** [comment]
- Docs and changelog have to be updated
- You don't handle the case when the user clicks back/forward browser buttons. In this case the tour should end, even when `allowUserInitiatedNavigation` is true
- Might be a good idea to update defaults

**Comment by mivanyna** [comment]
Yes, still working on it. Will ping you once it's done.

Which defaults? In docs?

**Comment by mivanyna** [comment]
@hakimio Do you have any good description for the property?

`Allow user navigate to other routes not ending the tour. False by default (optional)`

**Comment by hakimio** [comment]
> Which defaults? In docs?

https://github.com/hakimio/ngx-ui-tour/blob/6356be87bea7b1754a08ce90e132e41a0ff83d59/projects/ngx-ui-tour-core/src/lib/tour.service.ts#L60

**Comment by mivanyna** [comment]
I would even remove boolean properties with default value `false`.  Why keep them?

**Comment by hakimio** [comment]
I like explicitly defined defaults.

**Comment by mivanyna** [comment]
@hakimio please check

**Comment by mivanyna** [comment]
Should I update all the CHANGELOG files?

File: projects/ngx-ui-tour-console/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,10 @@
+<a name="10.4.1"></a>
+
+# [10.4.1](https://github.com/hakimio/ngx-ui-tour) (2023-03-07)
+
+### Features
+- Allow user initiated navigation when the tour is active.
```
Comment (hakimio) [inline projects/ngx-ui-tour-console/CHANGELOG.md:6]: > New `allowUserInitiatedNavigation` config. Enable to keep the tour open while the user navigates to a different page. Can cause the tour to get "stuck" if active anchor element is removed from the DOM after the navigation. The option is not related to navigation initiated by the UI tour and should not be enabled in most cases.

Also, new features increase minor version - not patch. [semver.org](https://semver.org/)

File: projects/ngx-ui-tour-core/src/lib/tour.service.ts [hunk 2]
```diff
@@ -158,8 +161,8 @@ export class TourService<T extends IStepOption = IStepOption> {
                 takeUntil(this.end$)
             )
             .subscribe(
-                () => {
-                    if (!this.navigationStarted) {
+                (event) => {
+                    if (!this.navigationStarted && (event.navigationTrigger === 'popstate' || !this.currentStep.allowUserInitiatedNavigation)) {
```
Comment (hakimio) [inline projects/ngx-ui-tour-core/src/lib/tour.service.ts:165]: Use variables to make the code more readable.
```ts
const browserBackBtnPressed = event.navigationTrigger === 'popstate',
    userNavigationAllowed = this.currentStep.allowUserInitiatedNavigation;

    if (!this.navigationStarted && (browserBackBtnPressed || !userNavigationAllowed) {}
```

File: src/app/shared/step-config/step-config.component.html [hunk 3]
```diff
@@ -199,4 +199,13 @@
         If your tour anchor is duplicated on purpose, you can choose to either register first or last anchor occurance
         instead of throwing an error.
     </ng-template>
+    <ng-template
+      documentationPropertyName="allowUserInitiatedNavigation"
+      documentationPropertyType="boolean"
+      [documentationPropertyValue]="false"
+    >
+        <div tourAnchor="config.allowUserInitiatedNavigation" style="cursor: pointer;">
+            Allow user navigate to other routes not ending the tour. False by default (optional)
```
Comment (hakimio) [inline src/app/shared/step-config/step-config.component.html:208]: > Enable to keep the tour open while the user navigates to a different page. Can cause the tour to get "stuck" if active anchor element is removed from the DOM after the navigation. The option is not related to navigation initiated by the UI tour and should not be enabled in most cases.

**Review by hakimio — CHANGES_REQUESTED** [review]

File: src/app/shared/step-config/step-config.component.html [hunk 4]
```diff
@@ -199,4 +199,15 @@
         If your tour anchor is duplicated on purpose, you can choose to either register first or last anchor occurance
         instead of throwing an error.
     </ng-template>
+    <ng-template
+      documentationPropertyName="allowUserInitiatedNavigation"
+      documentationPropertyType="boolean"
+      [documentationPropertyValue]="false"
+    >
+        <div tourAnchor="config.allowUserInitiatedNavigation" style="cursor: pointer;">
```
Comment (hakimio) [inline src/app/shared/step-config/step-config.component.html:207]: I don't think the `tourAnchor` is used.

**Review by hakimio — CHANGES_REQUESTED** [review]

File: src/app/shared/step-config/step-config.component.html (hunk 4 above)
Comment (hakimio) [reply src/app/shared/step-config/step-config.component.html:207]: And why does it need `style="cursor: pointer;"`?

File: projects/ngx-ui-tour-core/src/lib/tour.service.ts [hunk 5]
```diff
@@ -158,8 +161,11 @@ export class TourService<T extends IStepOption = IStepOption> {
                 takeUntil(this.end$)
             )
             .subscribe(
-                () => {
-                    if (!this.navigationStarted) {
+                (event) => {
+                    const browserBackBtnPressed = event.navigationTrigger === 'popstate',
+                      userNavigationAllowed = this.currentStep.allowUserInitiatedNavigation;
```
Comment (hakimio) [inline projects/ngx-ui-tour-core/src/lib/tour.service.ts:166]: Might be a good idea to use optional chaining in case `currentStep` is undefined.

**Review by hakimio — CHANGES_REQUESTED** [review]

**Pull request merged**